providers/implementations/keymgmt/ml_dsa_kmgmt.c

```diff
@@ -0,0 +1,375 @@
+        if (priv == NULL
+                || !OSSL_PARAM_set_octet_string(p, priv,
+                                                ossl_ml_dsa_key_get_priv_len(key)))
+            return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_PUB_KEY);
+    if (p != NULL) {
+        if (pub == NULL
+                || !OSSL_PARAM_set_octet_string(p, pub,
+                                                ossl_ml_dsa_key_get_pub_len(key)))
+            return 0;
+    }
+    return 1;
+}
+
+static int ml_dsa_export(void *keydata, int selection,
+                         OSSL_CALLBACK *param_cb, void *cbarg)
+{
+    ML_DSA_KEY *key = keydata;
+    OSSL_PARAM_BLD *tmpl;
+    OSSL_PARAM *params = NULL;
+    int ret = 0, include_private;
+
+    if (!ossl_prov_is_running() || key == NULL)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return 0;
+    /* The public key is required for private keys */
+    if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) == 0)
+        return 0;
+
+    tmpl = OSSL_PARAM_BLD_new();
+    if (tmpl == NULL)
+        return 0;
+
+    include_private = ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0);
+    if (!key_to_params(key, tmpl, include_private))
+        goto err;
+
+    params = OSSL_PARAM_BLD_to_param(tmpl);
+    if (params == NULL)
+        goto err;
+
+    ret = param_cb(params, cbarg);
+    OSSL_PARAM_free(params);
+err:
+    OSSL_PARAM_BLD_free(tmpl);
+    return ret;
+}
+
+static void *ml_dsa_load(const void *reference, size_t reference_sz)
+{
+    ML_DSA_KEY *key = NULL;
+
+    if (ossl_prov_is_running() && reference_sz == sizeof(key)) {
+        /* The contents of the reference is the address to our object */
+        key = *(ML_DSA_KEY **)reference;
+        /* We grabbed, so we detach it */
+        *(ML_DSA_KEY **)reference = NULL;
+        return key;
+    }
+    return NULL;
+}
+
+static void *ml_dsa_gen_init(void *provctx, int selection,
+                             const OSSL_PARAM params[])
+{
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(provctx);
+    struct ml_dsa_gen_ctx *gctx = NULL;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    if ((gctx = OPENSSL_zalloc(sizeof(*gctx))) != NULL) {
+        gctx->libctx = libctx;
+        if (!ml_dsa_gen_set_params(gctx, params)) {
+            OPENSSL_free(gctx);
+            gctx = NULL;
+        }
+    }
+    return gctx;
+}
+
+static void *ml_dsa_gen(void *genctx, const char *alg)
+{
+    struct ml_dsa_gen_ctx *gctx = genctx;
+    ML_DSA_KEY *key = NULL;
+    ML_DSA_CTX *ctx = NULL;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+    ctx = ossl_ml_dsa_ctx_new(alg, gctx->libctx, gctx->propq);
+    if (ctx == NULL)
+        return NULL;
+    key = ossl_ml_dsa_key_new(gctx->libctx, alg);
+    if (key == NULL)
+        return NULL;
+    if (!ossl_ml_dsa_generate_key(ctx, gctx->libctx,
+                                  gctx->entropy, gctx->entropy_len, key))
+        goto err;
```

Comment:
Perhaps, once key generation succeeds (or even fails), the provided entropy should be explicitly erased (that is, `OPENSSL_cleanse()`)? The seed is sensitive data.

Reply:
This is done later by ml_dsa_gen_cleanup()